spring-pulsar/src/main/java/org/springframework/pulsar/listener/DefaultPulsarMessageListenerContainer.java

```diff
@@ -282,145 +275,27 @@ else if (messageListener != null) {
-			}
-			if (properties.containsKey("patternAutoDiscoveryPeriod")) {
-				builder.patternAutoDiscoveryPeriod(
-						Integer.parseInt(properties.get("patternAutoDiscoveryPeriod").toString()), TimeUnit.SECONDS);
-			}
-			if (properties.containsKey("regexSubscriptionMode")) {
-				builder.subscriptionTopicsMode(
-						RegexSubscriptionMode.valueOf(properties.get("regexSubscriptionMode").toString()));
-			}
+
+			DeadLetterPolicy deadLetterPolicy = null;
 			if (properties.containsKey("deadLetterPolicy")) {
-				builder.deadLetterPolicy((DeadLetterPolicy) properties.get("deadLetterPolicy"));
-			}
-			if (properties.containsKey("retryEnable")) {
-				builder.enableRetry(Boolean.parseBoolean(properties.get("retryEnable").toString()));
-			}
-			if (properties.containsKey("batchReceivePolicy")) {
-				builder.batchReceivePolicy((BatchReceivePolicy) properties.get("batchReceivePolicy"));
-			}
-			if (properties.containsKey("autoUpdatePartitions")) {
-				builder.autoUpdatePartitions(Boolean.parseBoolean(properties.get("autoUpdatePartitions").toString()));
-			}
-			if (properties.containsKey("autoUpdatePartitionsIntervalSeconds")) {
-				builder.autoUpdatePartitionsInterval(
-						Integer.parseInt(properties.get("autoUpdatePartitionsIntervalSeconds").toString()),
-						TimeUnit.SECONDS);
+				deadLetterPolicy = (DeadLetterPolicy) properties.remove("deadLetterPolicy");
 			}
-			if (properties.containsKey("replicateSubscriptionState")) {
-				builder.replicateSubscriptionState(
-						Boolean.parseBoolean(properties.get("replicateSubscriptionState").toString()));
-			}
-			if (properties.containsKey("resetIncludeHead")) {
-				builder.startMessageIdInclusive();
-			}
-			if (properties.containsKey("keySharedPolicy")) {
-				builder.keySharedPolicy((KeySharedPolicy) properties.get("keySharedPolicy"));
-			}
-			if (properties.containsKey("batchIndexAckEnabled")) {
-				builder.enableBatchIndexAcknowledgment(
-						Boolean.parseBoolean(properties.get("batchIndexAckEnabled").toString()));
-			}
-			if (properties.containsKey("ackReceiptEnabled")) {
-				builder.isAckReceiptEnabled(Boolean.parseBoolean(properties.get("ackReceiptEnabled").toString()));
-			}
-			if (properties.containsKey("poolMessages")) {
-				builder.poolMessages(Boolean.parseBoolean(properties.get("poolMessages").toString()));
-			}
-			if (properties.containsKey("payloadProcessor")) {
-				builder.messagePayloadProcessor((MessagePayloadProcessor) properties.get("payloadProcessor"));
+
+			ConsumerConfigurationData<T> confData = ((ConsumerBuilderImpl<T>) builder).getConf();
+			DeadLetterPolicy deadLetterPolicyDefault = confData.getDeadLetterPolicy();
+			confData.setDeadLetterPolicy(null);
+
+			// Now that DLP has been stripped from properties it is safe to call
+			// builder.loadConf
+			builder.loadConf(properties);
+
+			// Prefer the DLP from the container properties over any default DLP set on
+			// the consumer factory.
+			if (deadLetterPolicy != null) {
```

Comment:
I think all `JsonIgnored` properties of `ConsumerConfigurationData` should also be reapplied bc they are reset to `null` during `loadConf`.
Maybe we could have a common utility method for DPCF and DPLMC.

Reply:
Ha! Yeh I dug into this last night and was thinking of doing both (util and extend the props). I had a comment queued up that looked like:

> A key takeaway that I just realized is that whatever properties we support in @PulsarListener via 
> https://github.com/spring-projects-experimental/spring-pulsar/blob/b4203bedc8dc2f7bd2783231f12a906a0bf5710b/spring-pulsar/src/main/java/org/springframework/pulsar/annotation/PulsarListenerAnnotationBeanPostProcessor.java#L365-L367
> that also happen to be one of the @JsonIgnore properties in CCD must also be handled here specially. 
> 
> That list is pretty much as follows so we are clear for now w/ the current fields we are handling. 
> 
> This caveat was already present prior to this PR though. Just saying out loud. 
> 
> @JsonIgnore
> private MessageListener<T> messageListener;
> 
>   @JsonIgnore
>   private ConsumerEventListener consumerEventListener;
> 
>   @JsonIgnore
>   private RedeliveryBackoff negativeAckRedeliveryBackoff;
> 
>   @JsonIgnore
>   private RedeliveryBackoff ackTimeoutRedeliveryBackoff;
> 
>   @JsonIgnore
>   private transient MessagePayloadProcessor payloadProcessor;
> 
>   @JsonIgnore
>   private BatchReceivePolicy batchReceivePolicy;
> 
> @JsonIgnore
> private DeadLetterPolicy deadLetterPolicy;

So if we were to do a common utility we would want to cover all these fields. However, because the ones that we are currently supporting in the PulsarListener are being handled I wanted to push that off as we have enough going on in this PR already. Of course, if one were to use this w/o the PL and pass in one of the unhandled properties it would not be set. 

In summary - I would recommend getting this PR merged and then iterate on it from there. Wdyt?

Reply:
OK. LGTM 👍